### PR TITLE
doc: refine enable s5 document

### PR DESCRIPTION
**Pull request by wuxyintel**
Lifecycle Manager will be refined in v2.7, this patch will
refine enable s5 document to align with the latest code.

Tracked-On: #6652

Signed-off-by: Xiangyang Wu <xiangyang.wu@intel.com>

### CODE REVIEW

File: doc/tutorials/enable_s5.rst [hunk 1]
```diff
@@ -31,131 +31,197 @@ The diagram below shows the overall architecture:
...
-     pm_vuart_node="-s 1:0,lpc -l com2,/run/acrn/life_mngr_"$vm_name
+   .. code-block:: none
 
-     acrn-dm -A -m $mem_size -s 0:0,hostbridge \
-       ...
-       $pm_channel \
-       $pm_by_vuart \
-       $pm_vuart_node \
-       ...
+      $ cd acrn-hypervisor
```
Comment (gvancuts) [inline doc/tutorials/enable_s5.rst:156]:
```suggestion
      cd acrn-hypervisor
```

Comment (gvancuts) [reply doc/tutorials/enable_s5.rst:156]: It's best to not include the prompt (e.g. `$` or `#`) so people can more easily copy-paste the instructions.

Can you remove them in all other instances in this PR?

File: doc/tutorials/enable_s5.rst [hunk 2]
```diff
@@ -31,131 +31,197 @@ The diagram below shows the overall architecture:
...
 *********
 
-The procedure for enabling S5 is specific to the particular OS:
-
-* For Linux (LaaG) or Windows (WaaG), include these lines in the launch script:
-
-  .. code-block:: bash
+1. Configure communication vUART for Service VM and User VMs:
+
+   Add these lines in the hypervisor scenario XML file manually:
```
Comment (gvancuts) [inline doc/tutorials/enable_s5.rst:90]: Is this really the only option for now? We want to avoid as much as possible that users edit these XML files directly.

File: doc/tutorials/enable_s5.rst [hunk 3]
```diff
@@ -31,131 +31,197 @@ The diagram below shows the overall architecture:
...
         $ make life_mngr
 
-      .. note:: If there is no ``x86_64-w64-mingw32-gcc`` compiler, you can run ``sudo apt install gcc-mingw-w64-x86-64``
-         on Ubuntu to install it.
+      .. note:: If there is no ``x86_64-w64-mingw32-gcc`` compiler, you can run
+               ``sudo apt install gcc-mingw-w64-x86-64`` on Ubuntu to install it.
+
+   #) Copy ``s5_trigger_win.py`` into the WaaG VM.
+
+      .. note:: If there is no ``python 3.8.10``, you can install it on the WaaG VM.
```
Comment (gvancuts) [inline doc/tutorials/enable_s5.rst:224]: We should rephrase this to indicate the user must install Python in his WaaG.

**Review by gvancuts — CHANGES_REQUESTED** [review]

File: doc/tutorials/enable_s5.rst (hunk 2 above)
Comment (wuxyintel) [reply doc/tutorials/enable_s5.rst:90]: Thanks! Good question. This manual XML file modification is only needed in v2.7 due to config tool GUI limitation. If config tool updates are done, this manual modification can be removed.

File: doc/tutorials/enable_s5.rst (hunk 1 above)
Comment (wuxyintel) [reply doc/tutorials/enable_s5.rst:156]: Thanks for your comments, will update this in the whole document.

File: doc/tutorials/enable_s5.rst (hunk 3 above)
Comment (wuxyintel) [reply doc/tutorials/enable_s5.rst:224]: Thanks, will update this.

File: doc/tutorials/enable_s5.rst (hunk 2 above)
Comment (dbkinder) [reply doc/tutorials/enable_s5.rst:90]: @nanlin @junjiemao1 We need to keep the configurator updated for these new options in v2.7, right?

Comment (NanlinXie) [reply doc/tutorials/enable_s5.rst:90]: @dbkinder , Since configurator is ongoing redesign, thus we would like to hold the big change for configurator until we have new design implementation.

**Comment by wuxyintel** [comment]
@gvancuts Thanks for your review. I have updated document according to your comments, can you help double check them, thanks!

File: doc/tutorials/enable_s5.rst [hunk 4]
```diff
@@ -31,135 +31,199 @@ The diagram below shows the overall architecture:
...
+         <target_vm_id>0</target_vm_id>
+         <target_uart_id>2</target_uart_id>
+      </legacy_vuart>
+      ...
+      /* VM3 */
+      ...
+
+  .. note:: These vUART is emulated in the hypervisor; expose the node as ``/dev/ttySn``.
+     For the first User VM, the communication vUART index should be 1. For other User VMs,
+     the vUART (index is 1) shoulbe be configured as invalid, the communication vUART index
```
Comment (gvancuts) [inline doc/tutorials/enable_s5.rst:146]: I'm not able to understand this note. Can you rephrase/clarify?

File: doc/tutorials/enable_s5.rst [hunk 5]
```diff
@@ -31,135 +31,199 @@ The diagram below shows the overall architecture:
...
+     should be 2 or others.
+  .. note:: serial.conf file will be generated for Service VM if above communication vUARTs
+     are configured, need to copy ``build/hypervisor/serial.conf`` into ``/etc`` folder of Service VM.
```
Comment (gvancuts) [inline doc/tutorials/enable_s5.rst:149]: I'm assuming this would be done as part of the packaging or via the `make install` target, wouldn't it?

File: doc/tutorials/enable_s5.rst [hunk 6]
```diff
@@ -31,135 +31,199 @@ The diagram below shows the overall architecture:
...
+  .. note:: Need to install ``setserial`` tool in Service VM.
```
Comment (gvancuts) [inline doc/tutorials/enable_s5.rst:150]: This needs to be added to the GSG.

File: doc/tutorials/enable_s5.rst [hunk 7]
```diff
@@ -31,135 +31,199 @@ The diagram below shows the overall architecture:
...
-         $ cd acrn-hypervisor
-         $ make life_mngr
+         VM_TYPE=service_vm
+         VM_NAME=Service_VM
+         DEV_NAME=tty:/dev/ttyS8,/dev/ttyS9,/dev/ttyS10,/dev/ttyS11,/dev/ttyS12,/dev/ttyS13,/dev/ttyS14
+         ALLOW_TRIGGER_S5=/dev/ttySn
 
-   #. Copy ``life_mngr`` and ``life_mngr.service`` into the User VM:
+      .. note:: The mapping between User VM ID and communication serail device name (/dev/ttySn)
+                  in the /etc/serial.conf. If /dev/ttySn is configured in the  ``ALLOW_TRIGGER_S5``,
```
Comment (gvancuts) [inline doc/tutorials/enable_s5.rst:185]:
```suggestion
                  in the /file:`etc/serial.conf`. If ``/dev/ttySn`` is configured in the  ``ALLOW_TRIGGER_S5``,
```

File: doc/tutorials/enable_s5.rst [hunk 8]
```diff
@@ -31,135 +31,199 @@ The diagram below shows the overall architecture:
...
+                  this means system shutdown is allowed to be triggered in the the corresponding User VM.
```
Comment (gvancuts) [inline doc/tutorials/enable_s5.rst:186]:
```suggestion
                  this means system shutdown is allowed to be triggered in the corresponding User VM.
```

File: doc/tutorials/enable_s5.rst [hunk 9]
```diff
@@ -31,135 +31,199 @@ The diagram below shows the overall architecture:
...
+
+   #. Edit options in ``/etc/life_mngr/life_mngr.conf`` in the User VM.
 
       .. code-block:: none
 
-         $ scp build/misc/services/life_mngr root@<test board address>:/usr/bin/life_mngr
-         $ scp build/misc/services/life_mngr.service root@<test board address>:/lib/systemd/system/life_mngr.service
+         VM_TYPE=user_vm
```
Comment (gvancuts) [inline doc/tutorials/enable_s5.rst:192]: Do we describe anywhere what the different values for this field can be?

File: doc/tutorials/enable_s5.rst [hunk 10]
```diff
@@ -204,72 +268,69 @@ How to Test
...
+#. The Lifecycle Manager in the Service VM sends ``poweroff_cmd`` request to
+   Linux User VM.
+#. After receiving the ``poweroff_cmd`` request, the Lifecycle Manager in the
+   Linux User VM responds with an ack message, then shuts down VM.
+#. The Lifecycle Manager in the Service VM sends ``poweroff_cmd`` request to
+   Pre-launched RTVM.
+#. After receiving the ``poweroff_cmd`` request, the Lifecycle Manager in
+   the Pre-launched RTVM responds with an ack message.
+#. The Lifecycle Manager in the Pre-launched RTVM shuts down the VM using
+   standard PM registers.
```
Comment (gvancuts) [inline doc/tutorials/enable_s5.rst:333]: Can we rephrase this somehow? What does it mean to "use the standard PM registers", is this a traditional shutdown as if you were running on bare-metal hardware?

File: doc/tutorials/enable_s5.rst [hunk 11]
```diff
@@ -31,135 +31,199 @@ The diagram below shows the overall architecture:
...
-        $ make life_mngr
+        cd acrn-hypervisor
+        make life_mngr
 
-      .. note:: If there is no ``x86_64-w64-mingw32-gcc`` compiler, you can run ``sudo apt install gcc-mingw-w64-x86-64``
-         on Ubuntu to install it.
+      .. note:: If there is no ``x86_64-w64-mingw32-gcc`` compiler, you can run
+               ``sudo apt install gcc-mingw-w64-x86-64`` on Ubuntu to install it.
+
+   #) Copy ``s5_trigger_win.py`` into the WaaG VM, install python3 in the Waag.
```
Comment (gvancuts) [inline doc/tutorials/enable_s5.rst:222]: Does it need anything more than just Python3? Any pointer we can give for an installation guide on Windows?

**Review by gvancuts — CHANGES_REQUESTED** [review]

File: doc/tutorials/enable_s5.rst [hunk 12]
```diff
@@ -31,135 +31,199 @@ The diagram below shows the overall architecture:
...
+         <irq>COM2_IRQ</irq>
+         <target_vm_id>0</target_vm_id>
+         <target_uart_id>2</target_uart_id>
+      </legacy_vuart>
+      ...
+      /* VM3 */
+      ...
+
+  .. note:: These vUART is emulated in the hypervisor; expose the node as ``/dev/ttySn``.
+     For the first User VM, the communication vUART index should be 1. For other User VMs,
```
[the remark this answers is not in the document]
Comment (wuxyintel) [reply doc/tutorials/enable_s5.rst:145]: Thanks! This is the config tool limitation, vuart configuration in config tool may needs to be refined after v2.7. @yuchuyang Can you help to explain more detail about vUART configuration, thanks!

File: doc/tutorials/enable_s5.rst (hunk 4 above)
Comment (wuxyintel) [reply doc/tutorials/enable_s5.rst:146]: Thanks, will refine this.

File: doc/tutorials/enable_s5.rst (hunk 5 above)
Comment (wuxyintel) [reply doc/tutorials/enable_s5.rst:149]: Thanks! In current makefile, will install in the "$(DESTDIR)$(libdir)/acrn/", for ubuntu service VM, user can use deb package install this file into "/etc", for yocto service VM, need to copy this file into /etc manually.
If install this file in /etc directly in makefile, the yocto build will fail.

File: doc/tutorials/enable_s5.rst (hunk 7 above)
Comment (wuxyintel) [reply doc/tutorials/enable_s5.rst:185]: Thanks, will update this.

File: doc/tutorials/enable_s5.rst (hunk 8 above)
Comment (wuxyintel) [reply doc/tutorials/enable_s5.rst:186]: Thanks, will update this.

File: doc/tutorials/enable_s5.rst (hunk 9 above)
Comment (wuxyintel) [reply doc/tutorials/enable_s5.rst:192]: Thanks! More details are described in the "/etc/life_mngr/life_mngr.conf".

File: doc/tutorials/enable_s5.rst (hunk 10 above)
Comment (wuxyintel) [reply doc/tutorials/enable_s5.rst:333]: Thanks, it is ACPI PM register, will update this.

File: doc/tutorials/enable_s5.rst (hunk 11 above)
Comment (wuxyintel) [reply doc/tutorials/enable_s5.rst:222]: Thanks, will add download and install guide on Windows, thanks! Only need python3

File: doc/tutorials/enable_s5.rst (hunk 6 above)
Comment (wuxyintel) [reply doc/tutorials/enable_s5.rst:150]: Thanks, will move this into "Install ACRN" part of GSG.

File: doc/tutorials/enable_s5.rst [hunk 13]
```diff
@@ -9,8 +9,8 @@ Introduction
 S5 is one of the `ACPI sleep states <http://acpi.sourceforge.net/documentation/sleep.html>`_
 that refers to the system being shut down (although some power may still be
 supplied to certain devices). In this document, S5 means the function to
-shut down the **User VMs**, **the Service VM**, the hypervisor, and the
-hardware. In most cases, directly shutting down the power of a computer
+shut down the **User VMs**, **Service VM**, the hypervisor, and the
+hardware. In most cases, directly shutting down the power off a computer
```
[the remark this answers is not in the document]
Comment (wuxyintel) [reply doc/tutorials/enable_s5.rst:13]: Thanks, will update this.

File: doc/getting-started/getting-started.rst [hunk 14]
```diff
@@ -576,7 +576,13 @@ Build ACRN
 Install ACRN
 ************
 
-In the following steps, you will configure GRUB on the target system.
+In the following steps, you will install serial configuration tool and configure GRUB on the target system.
+
+#. Install serial configurtion tool in the target system as follows:
```
Comment (gvancuts) [inline doc/getting-started/getting-started.rst:581]:
```suggestion
#. Install serial configuration tool in the target system as follows:
```

File: doc/tutorials/enable_s5.rst [hunk 15]
```diff
@@ -31,135 +31,202 @@ The diagram below shows the overall architecture:
...
+         <base>COM2_BASE</base>
+         <irq>COM2_IRQ</irq>
+         <target_vm_id>0</target_vm_id>
+         <target_uart_id>2</target_uart_id>
+      </legacy_vuart>
+      ...
+      /* VM3 */
+      ...
+
+  .. note:: These vUART is emulated in the hypervisor; expose the node as ``/dev/ttySn``.
```
Comment (gvancuts) [inline doc/tutorials/enable_s5.rst:144]: I do not understand this restriction. Is `legacy_vuart_id` a global number, or is it local to a VM definition?

File: doc/tutorials/enable_s5.rst (hunk 5 above)
Comment (gvancuts) [reply doc/tutorials/enable_s5.rst:149]: Interesting, do we know why the Yocto build fails if we copy the file to `/etc`?

So only the Debian packaging will actually install this file in the correct location, am I understanding this correctly?

File: doc/tutorials/enable_s5.rst [hunk 16]
```diff
@@ -31,135 +31,202 @@ The diagram below shows the overall architecture:
...
       .. code-block:: none
 
-         $ cd acrn-hypervisor
-         $ make life_mngr
+         VM_TYPE=service_vm
+         VM_NAME=Service_VM
+         DEV_NAME=tty:/dev/ttyS8,/dev/ttyS9,/dev/ttyS10,/dev/ttyS11,/dev/ttyS12,/dev/ttyS13,/dev/ttyS14
+         ALLOW_TRIGGER_S5=/dev/ttySn
+
+      .. note:: The mapping between User VM ID and communication serail device name (/dev/ttySn)
```
Comment (gvancuts) [inline doc/tutorials/enable_s5.rst:184]:
```suggestion
      .. note:: The mapping between User VM ID and communication serial device name (``/dev/ttySn``)
```

File: doc/tutorials/enable_s5.rst [hunk 17]
```diff
@@ -31,135 +31,202 @@ The diagram below shows the overall architecture:
...
+                  in the file: `/etc/serial.conf`. If ``/dev/ttySn`` is configured in the  ``ALLOW_TRIGGER_S5``,
```
Comment (gvancuts) [inline doc/tutorials/enable_s5.rst:185]:
```suggestion
                  in the :file:`/etc/serial.conf`. If ``/dev/ttySn`` is configured in the  ``ALLOW_TRIGGER_S5``,
```

File: doc/tutorials/enable_s5.rst [hunk 18]
```diff
@@ -31,135 +31,202 @@ The diagram below shows the overall architecture:
...
+      </legacy_vuart>
+      ...
+      /* VM3 */
+      ...
+
+  .. note:: These vUART is emulated in the hypervisor; expose the node as ``/dev/ttySn``.
+     For the User VM with the minimal VM ID, the communication vUART id should be 1.
+     For other User VMs, the vUART (id is 1) shoulbe be configured as invalid, the communication
+     vUART id should be 2 or others.
+  .. note:: serial.conf file will be generated for Service VM if above communication vUARTs
```
Comment (gvancuts) [inline doc/tutorials/enable_s5.rst:148]: What is in this `serial.conf` file?

File: doc/tutorials/enable_s5.rst [hunk 19]
```diff
@@ -31,135 +31,202 @@ The diagram below shows the overall architecture:
...
-     pm_channel="--pm_notify_channel uart"
-     pm_by_vuart="--pm_by_vuart tty,/dev/ttyS1"
+      .. code-block:: none
 
-     /usr/bin/acrn-dm -A -m $mem_size -s 0:0,hostbridge \
-        ...
-        $pm_channel \
-        $pm_by_vuart \
-        ...
+         scp build/misc/services/s5_trigger_linux.py root@<target board address>:~/
```
Comment (gvancuts) [inline doc/tutorials/enable_s5.rst:166]: Wouldn't it be a good place here to also copy the `serial.conf` file?

File: doc/tutorials/enable_s5.rst [hunk 20]
```diff
@@ -31,135 +31,202 @@ The diagram below shows the overall architecture:
...
-         on Ubuntu to install it.
+   #) Copy ``s5_trigger_win.py`` into the WaaG VM.
 
    #) Set up a Windows environment:
 
-      I) Download the :kbd:`Visual Studio 2019` tool from `<https://visualstudio.microsoft.com/downloads/>`_,
+      1) Download the Python3 from `<https://www.python.org/downloads/release/python-3810/>`_, install
+         "Python 3.8.10" in WaaG.
+
+      #) Download the Visual Studio 2019 tool from `<https://visualstudio.microsoft.com/downloads/>`_,
```
Comment (gvancuts) [inline doc/tutorials/enable_s5.rst:229]: What do we need the Visual Studio tool for? This seems like a serious constraint for people who want to deploy a WaaG VM and enable S5 in it.

File: doc/tutorials/enable_s5.rst [hunk 21]
```diff
@@ -204,72 +271,69 @@ How to Test
...
-#. Use the ``acrnctl stop`` command on the Service VM to trigger S5 to the User VMs:
+#. Use the ``user_vm_shutdown.py`` in the Service VM to shut down the User VMs:
+
+   .. code-block:: none
 
-   .. code-block:: console
+      sudo python3 ~/user_vm_shutdown.py <User VM name>
 
-      # acrnctl stop vm1
+   .. note:: The User VM name is configured in the life_mngr.conf of User VM.
```
Comment (gvancuts) [inline doc/tutorials/enable_s5.rst:292]:
```suggestion
   .. note:: The User VM name is configured in the :file:`life_mngr.conf` of User VM.
```

File: doc/tutorials/enable_s5.rst [hunk 22]
```diff
@@ -204,72 +271,69 @@ How to Test
...
 
 #. Use the ``acrnctl list`` command to check the User VM status.
 
-   .. code-block:: console
+   .. code-block:: none
 
-      # acrnctl list
-      vm1		stopped
+      acrnctl list
```
Comment (gvancuts) [inline doc/tutorials/enable_s5.rst:298]: Assuming you need to be root for this (I'm not 100% sure anymore but believe it to be the case), add `sudo`

File: doc/getting-started/getting-started.rst (hunk 14 above)
Comment (wuxyintel) [reply doc/getting-started/getting-started.rst:581]: Thanks, will update this.

File: doc/tutorials/enable_s5.rst (hunk 15 above)
Comment (wuxyintel) [reply doc/tutorials/enable_s5.rst:144]: Maybe legacy vuart configuration is set by config tool in hard code way, I think this is temporary solution. @NanlinXie Can you help to explain more about these vuart configuration. Thanks!

File: doc/tutorials/enable_s5.rst (hunk 16 above)
Comment (wuxyintel) [reply doc/tutorials/enable_s5.rst:184]: For yocto, I have tried to install /etc, but get permission failure. This is similar acrn.bin, we also need copy acrn.bin into /boot folder manually. BTW, I will put copy serial.conf in to GSG.

File: doc/tutorials/enable_s5.rst (hunk 17 above)
Comment (wuxyintel) [reply doc/tutorials/enable_s5.rst:185]: This file is generated during ACRN hypervisor build, not lifecycle manager build, so I will put copy serial.conf in to GSG.

File: doc/tutorials/enable_s5.rst (hunk 18 above)
Comment (wuxyintel) [reply doc/tutorials/enable_s5.rst:148]: serial.conf is to help service VM to configure vUART emulated by hypervisor. It is generated during ACRN hypervisor build.

File: doc/tutorials/enable_s5.rst (hunk 21 above)
Comment (wuxyintel) [reply doc/tutorials/enable_s5.rst:292]: Thanks, will update this.

File: doc/tutorials/enable_s5.rst (hunk 22 above)
Comment (wuxyintel) [reply doc/tutorials/enable_s5.rst:298]: Thanks, you are right, will add sudo.

File: doc/tutorials/enable_s5.rst (hunk 20 above)
Comment (wuxyintel) [reply doc/tutorials/enable_s5.rst:229]: If Lifecycle Manager for WaaG will be built in Windows, need to install Visual Studio tool. If Lifecycle Manager for WaaG is built in Linux, this tool is not needed. BTW, I will update document about this.

**Review by gvancuts — APPROVED** [review]
